src/account/sync/mod.rs

```diff
@@ -2748,13 +2754,52 @@ pub(crate) async fn repost_message(
     Ok(message)
 }
 
+fn verify_unlock_blocks(
+    transaction_payload: &TransactionPayload,
+    mut inputs: Vec<(Input, BeeAddress)>,
+) -> crate::Result<()> {
+    // Sort inputs
+    inputs.sort_by(|a, b| a.0.cmp(&b.0));
+    let essence_hash = transaction_payload.essence().hash();
+    let unlock_blocks = transaction_payload.unlock_blocks();
+    for (index, (_input, address)) in inputs.iter().enumerate() {
+        verify_signature(address, unlock_blocks, index, &essence_hash)?;
+    }
+    Ok(())
+}
+
+fn verify_signature(
+    address: &BeeAddress,
+    unlock_blocks: &UnlockBlocks,
+    index: usize,
+    essence_hash: &[u8; 32],
+) -> crate::Result<()> {
+    let signature_unlock_block = match unlock_blocks.get(index) {
+        Some(unlock_block) => match unlock_block {
+            UnlockBlock::Signature(b) => b,
+            UnlockBlock::Reference(b) => match unlock_blocks.get(b.index().into()) {
```

Comment:
I think get is already resolving the reference ?

Reply:
Sure but isn't that what `UnlockBlocks::get` does in the first place ?

Reply:
Ah it does it internally, will update